apollo-router/src/uplink/entitlement_stream.rs

```diff
@@ -0,0 +1,386 @@
+    struct EntitlementExpander<Upstream>
+    where
+        Upstream: Stream<Item = Event>,
+    {
+        #[pin]
+        checks: DelayQueue<Event>,
+        #[pin]
+        upstream: Upstream,
+    }
+}
+
+impl<Upstream> Stream for EntitlementExpander<Upstream>
+where
+    Upstream: Stream<Item = Event>,
+{
+    type Item = Event;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        let active_checks = !this.checks.is_empty();
+        match this.checks.poll_expired(cx) {
+            // We have an expired claim that needs checking
+            Poll::Ready(Some(item)) => Poll::Ready(Some(item.into_inner())),
+
+            // No expired claim is ready so go for upstream
+            Poll::Pending | Poll::Ready(None) => {
+                let next = this.upstream.poll_next(cx);
+                match (&next, active_checks) {
+                    // Upstream has a new event with a claim
+                    (
+                        Poll::Ready(Some(Event::UpdateEntitlement(Entitlement {
+                            claims: Some(claim),
+                            ..
+                        }))),
+                        _,
+                    ) => {
+                        // We got a new claim, so clear the previous checks.
+                        this.checks.clear();
+                        // Insert the new checks.
+                        this.checks
+                            .insert_at(Event::WarnEntitlement, (to_instant(claim.warn_at)).into());
+                        this.checks
+                            .insert_at(Event::HaltEntitlement, (to_instant(claim.halt_at)).into());
+                        next
+                    }
+                    // Upstream has a new event with no claim
+                    (Poll::Ready(Some(_)), _) => {
+                        // As we have no claim clear the checks
+                        this.checks.clear();
+                        next
+                    }
+                    // There were still active checks, so go for pending
+                    (Poll::Ready(None), true) => Poll::Pending,
+                    // Upstream is exhausted and there are no checks left
+                    (Poll::Ready(None), false) => Poll::Ready(None),
+                    // Upstream not exhausted
+                    (Poll::Pending, _) => Poll::Pending,
+                }
+            }
+        }
+    }
+}
+
+fn to_instant(system_time: SystemTime) -> Instant {
+    // This is approximate as there is no real conversion between SystemTime and Instant
+    let now_instant = Instant::now();
+    let now_system_time = SystemTime::now();
+    match now_system_time.duration_since(system_time) {
+        Ok(duration) => now_instant
+            .checked_sub(duration)
+            .expect("system time can always be converted into an instant"),
+        Err(e) => now_instant + e.duration(),
```

Comment:
Is this calculation correct?

So: If we succeed we get now - (now_system_time - system_time), if we fail we get now + (system_time - now_system_time).

I'm not clear how these are equivalent. In the case of a failure, the value of e.duration() could be almost anything couldn't it?

BTW: I'm not saying the calculation is wrong, but if it's correct, it could certainly do with a comment describing exactly what it is calculating.

Reply:
Feedback in f864aa6